### ISSUE TITLE
[GI] utf8 string annotations

### ISSUE THREAD

**Issue by behdad** [harfbuzz/harfbuzz#4157]
Any reason why we annotate our UTF-8 strings as:
```
@str: (array length=len) (element-type uint8_t):
```
I have copy/pasted that from our first instance. Problem is, now testing it, it doesn't work from Python. Whereas if I remove the `element-type` then it works for a Python string.

**Comment by matthiasclasen** [comment]
I believe utf8 strings should not have an element-type annotation.

**Comment by ebassi** [comment]
Correct. C strings should either be without annotation, to signal they are UTF-8 encoded; or should use a `filename` annotation, to signal that they are Unix paths, using whatever native file system encoding.

C strings should **definitely** never be annotated as arrays of characters. They should also **never** have a `length` argument as a way to pass a length in bytes or, worse, `-1` to let the callee do the `strlen()`. Any API that takes a UTF-8 string in the form of:

```c
void
foo_set_string (Foo *self,
                const char *str,
                ssize_t len);
```

is automatically a broken API, regardless of any annotation. Drop the `len`, and mandate zero-terminated UTF-8 C strings.

The only type that should be marked as `array` (zero-terminated, fixed length, or parametrised length) is a bytes array, but in that case:

- it may not be a NUL-terminated UTF-8 encoded C string
- it should use `(uint8_t *data, size_t len)` as the arguments
